jax/experimental/lanczos.py

```diff
@@ -0,0 +1,308 @@
+                                - alpha * Q[:, :-1]
+                                - beta * Q[:, k][:, jnp.newaxis])
+      )
+      precondition = abs(should_be_zero).sum(axis=0)
+    Q, alpha, beta = _lanczos_restart(A, k, m, Q, alpha, beta)
+
+    T = _build_arrowhead_matrix(alpha, beta, k)
+    ritz_values, Y = jnp.linalg.eigh(T)
+
+    residual_norm = beta[-1] * abs(Y[-1, :])
+    converged = residual_norm < tolerance
+    num_converged = jnp.where(
+        converged.all(), inner_iterations, jnp.argmin(converged)
+    )
+
+    saved, gamma = _pick_saved_ritz_values(
+        ritz_values, num_converged, num_desired)
+    k_next = jnp.sum(saved)
+    saved_to_front = jnp.argsort(~saved)
+    ritz_values = ritz_values[saved_to_front]
+    Y = Y[:, saved_to_front]
+
+    updated = k_next > jnp.arange(m)
+    alpha_hat = jnp.where(updated, ritz_values, jnp.zeros_like(alpha))
+    beta_hat = jnp.where(updated, beta[-1] * Y[-1, :], jnp.zeros_like(beta))
+    Q_hat = Q.at[:, :-1].set(jnp.where(updated, _dot(Q[:, :-1], Y), Q[:, :-1]))
+    Q_hat = Q_hat.at[:, k_next].set(Q[:, -1])
+
+    if return_info:
+      info = _tree_index_update(info, restart, {
+          'k': k,
+          'num_converged': num_converged,
+          'residual_norm': residual_norm,
+          'alpha': alpha,
+          'beta': beta,
+          'alpha_hat': alpha_hat,
+          'beta_hat': beta_hat,
+          'converged': converged,
+          'saved': saved,
+          'gamma': gamma,
+          'precondition': precondition,
+      })
+
+    return _ThickRestartState(
+        Q_hat, alpha_hat, beta_hat, restart + 1, k_next, num_converged, info)
+
+  q0 = q0 / jnp.linalg.norm(q0)
+  n = q0.shape[0]
+
+  Q = jnp.zeros((n, m+1)).at[:, 0].set(q0)
+  alpha = jnp.zeros(m)
+  beta = jnp.zeros(m)
+  if return_info:
+    info = {
+        'k': jnp.zeros((max_restarts,), dtype=jnp.int32),
+        'num_converged': jnp.zeros((max_restarts,), dtype=jnp.int32),
+        'residual_norm': jnp.zeros((max_restarts, inner_iterations)),
+        'alpha': jnp.zeros((max_restarts, inner_iterations)),
+        'beta': jnp.zeros((max_restarts, inner_iterations)),
+        'alpha_hat': jnp.zeros((max_restarts, inner_iterations)),
+        'beta_hat': jnp.zeros((max_restarts, inner_iterations)),
+        'converged': jnp.zeros((max_restarts, inner_iterations), dtype=jnp.int32),
+        'saved': jnp.zeros((max_restarts, inner_iterations), dtype=jnp.int32),
+        'gamma': jnp.zeros((max_restarts, inner_iterations)),
+        'precondition': jnp.zeros((max_restarts, inner_iterations)),
+    }
+  else:
+    info = None
+
+  return lax.while_loop(
+      cond_fun, body_fun, _ThickRestartState(Q, alpha, beta, 0, 0, 0, info)
+  )
+
+
+def eigsh_smallest(
+    A,
+    q0,
+    num_desired,
+    inner_iterations=None,
+    max_restarts=None,
+    tolerance=None,
+    return_info=False,
+):
+  """Find the `num_desired` smallest eigenvalues of the linear map A.
```

Comment:
My experience with lanczos solvers is that they perform poorly for finding the smallest eigenvalues except in shift-invert mode, in which the smallest eigenvalues are transformed to the largest.

Have you tried modifying this to find the largest eigenvalues?